syntax = "proto3";

option java_multiple_files = true;
option java_package = "io.peripher.proto";
option java_outer_classname = "PeripherioProto";
option objc_class_prefix = "PPI";

package peripherio;

service Peripherio {
  /* Device management calls */
  rpc List(Config) returns (FindResponse) {}
  rpc Find(FindRequest) returns (FindResponse) {}
  rpc PingDevice(PingRequest) returns (PingResponse) {}

  /* Driver management calls */
  rpc FindDrivers(FindRequest) returns (FindDriversResponse) {}

  /* Command Dispatchment calls */
  rpc Dispatch(DispatchRequest) returns (DispatchResponse) {}
}

message DispatchRequest {
  DeviceID device = 1;
  string command = 2;
  bytes args = 3; // msgpack
}

message DispatchResponse {
  bytes rets = 1; // msgpack
}

message PingResponse {
  bool alive = 1;
}

message FindRequest {
  Config config = 1;
  DriverSpecification spec = 2;
}

message PingRequest {
  Config config = 1;
  DeviceID device = 2;
}

message FindResponse {
  message DetectResult {
    DeviceID id = 1;
    string display_name = 2;
    Config config = 3;
  }
  repeated DetectResult results = 1;
}

message Config {
  message Pair {
    string key = 1;
    bytes value = 2; // msgpack
  }
  repeated Pair config = 1;
}

message DeviceID {
  uint64 id = 1;
}

message DriverSpecification {
  string category = 1;
  string vendor = 2;
  string name = 3;
}

message Driver {
  string name = 1;
  string version = 2;
  string vendor = 3;
  string path = 4;
  repeated string category = 5;
}

message FindDriversResponse {
  repeated Driver results = 1;
}
